lib/events/firestoreevents/firestoreevents.go

```diff
@@ -400,51 +400,31 @@ func (l *Log) SearchEvents(fromUTC, toUTC time.Time, namespace string, eventType
 	return l.searchEventsWithFilter(fromUTC, toUTC, namespace, limit, order, startKey, searchEventsFilter{eventTypes: eventTypes})
 }
 
-func (l *Log) searchEventsWithFilter(fromUTC, toUTC time.Time, namespace string, limit int, order types.EventOrder, startKey string, filter searchEventsFilter) ([]apievents.AuditEvent, string, error) {
-	var eventsArr []apievents.AuditEvent
-	var estimatedSize int
-	checkpoint := startKey
-	left := limit
+func (l *Log) searchEventsWithFilter(fromUTC, toUTC time.Time, namespace string, limit int, order types.EventOrder, lastKey string, filter searchEventsFilter) ([]apievents.AuditEvent, string, error) {
+	g := l.WithFields(log.Fields{"From": fromUTC, "To": toUTC, "Namespace": namespace, "Filter": filter, "Limit": limit, "StartKey": lastKey})
 
-	for {
-		gotEvents, withSize, withCheckpoint, err := l.searchEventsOnce(fromUTC, toUTC, namespace, left, order, checkpoint, filter, events.MaxEventBytesInResponse-estimatedSize)
-		if nil != err {
-			return nil, "", trace.Wrap(err)
-		}
-
-		eventsArr = append(eventsArr, gotEvents...)
-		estimatedSize += withSize
-		left -= len(gotEvents)
-		checkpoint = withCheckpoint
-
-		if len(checkpoint) == 0 || left <= 0 || estimatedSize >= events.MaxEventBytesInResponse {
-			break
-		}
-	}
-
-	return eventsArr, checkpoint, nil
-}
-
-func (l *Log) searchEventsOnce(fromUTC, toUTC time.Time, namespace string, limit int, order types.EventOrder, startKey string, filter searchEventsFilter, spaceRemaining int) ([]apievents.AuditEvent, int, string, error) {
-	g := l.WithFields(log.Fields{"From": fromUTC, "To": toUTC, "Namespace": namespace, "Filter": filter, "Limit": limit, "StartKey": startKey})
-
-	var lastKey int64
 	var values []events.EventFields
-	var parsedStartKey int64
 	var err error
-	reachedEnd := false
 	totalSize := 0
+	var checkpointParts []string
+	var checkpointTime int
 
-	if startKey != "" {
-		parsedStartKey, err = strconv.ParseInt(startKey, 10, 64)
+	if lastKey != "" {
+		checkpointParts = strings.Split(lastKey, ":")
+		checkpointTime, err = strconv.Atoi(checkpointParts[0])
 		if err != nil {
-			return nil, 0, "", trace.WrapWithMessage(err, "failed to parse startKey, expected integer but found: %q", startKey)
+			return nil, "", trace.BadParameter("invalid checkpoint key: %q", lastKey)
 		}
 	}
 
 	modifyquery := func(query firestore.Query) firestore.Query {
-		if startKey != "" {
-			return query.StartAfter(parsedStartKey)
+		if len(filter.eventTypes) > 0 {
+			query = query.Where(eventTypeDocProperty, "in", filter.eventTypes)
+		}
+
+		if lastKey != "" {
+			query = query.OrderBy(firestore.DocumentID, firestore.Asc)
```

Comment:
What's this added `OrderBy` for? Wouldn't this change the order of the pages after retrieving the initial page?

Reply:
Ah, good catch. This is needed because every StartAfter field must be matched by an OrderBy clause. I've added it to every query now regardless of pagination key.